x/marketmaker/keeper/marketmaker.go

```diff
@@ -0,0 +1,356 @@
+		if _, ok := incentivePairsMap[pairId]; !ok {
+			return types.ErrUnregisteredPairId
+		}
+
+	}
+
+	// deposit deposit amount * number of pairs
+	err := k.bankKeeper.SendCoinsFromAccountToModule(ctx, mmAddr, types.ModuleName, totalDepositAmt)
+	if err != nil {
+		return err
+	}
+
+	// create deposit object for refund when inclusion or exclusion
+	for _, pairId := range pairIds {
+		k.SetDeposit(ctx, mmAddr, pairId, params.DepositAmount)
+	}
+
+	// create market maker object for each pair
+	for _, pairId := range pairIds {
+		k.SetMarketMaker(ctx, types.MarketMaker{
+			Address:  mmAddr.String(),
+			PairId:   pairId,
+			Eligible: false,
+		})
+	}
+
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			types.EventTypeApplyMarketMaker,
+			sdk.NewAttribute(types.AttributeKeyAddress, mmAddr.String()),
+			sdk.NewAttribute(types.AttributeKeyPairIds, strings.Trim(strings.Replace(fmt.Sprint(pairIds), " ", ",", -1), "[]")),
+		),
+		sdk.NewEvent(
+			sdk.EventTypeMessage,
+			sdk.NewAttribute(sdk.AttributeKeyModule, types.AttributeValueCategory),
+			sdk.NewAttribute(sdk.AttributeKeyAction, types.EventTypeApplyMarketMaker),
+			sdk.NewAttribute(sdk.AttributeKeySender, mmAddr.String()),
+		),
+	})
+
+	return nil
+}
+
+func (k Keeper) ClaimIncentives(ctx sdk.Context, mmAddr sdk.AccAddress) error {
+	incentive, found := k.GetIncentive(ctx, mmAddr)
+	if !found {
+		return types.ErrEmptyClaimableIncentive
+	}
+
+	if err := k.bankKeeper.SendCoins(ctx, types.ClaimableIncentiveReserveAcc, mmAddr, incentive.Claimable); err != nil {
+		return err
+	}
+
+	k.DeleteIncentive(ctx, mmAddr)
+
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			types.EventTypeClaimIncentives,
+			sdk.NewAttribute(types.AttributeKeyAddress, mmAddr.String()),
+			sdk.NewAttribute(sdk.AttributeKeyAmount, incentive.Claimable.String()),
+		),
+		sdk.NewEvent(
+			sdk.EventTypeMessage,
+			sdk.NewAttribute(sdk.AttributeKeyModule, types.AttributeValueCategory),
+			sdk.NewAttribute(sdk.AttributeKeyAction, types.EventTypeClaimIncentives),
+			sdk.NewAttribute(sdk.AttributeKeySender, mmAddr.String()),
+		),
+	})
+	return nil
+}
+
+func (k Keeper) ValidateDepositReservedAmount(ctx sdk.Context) error {
+	mmCount := 0
+	depositCount := 0
+	var totalAmt sdk.Coins
+	k.IterateMarketMakers(ctx, func(mm types.MarketMaker) (stop bool) {
+		if !mm.Eligible {
+			mmCount += 1
+		}
+		return false
+	})
+	k.IterateDeposits(ctx, func(id types.Deposit) (stop bool) {
+		depositCount += 1
+		totalAmt = totalAmt.Add(id.Amount...)
+		return false
+	})
+	if mmCount != depositCount {
+		return fmt.Errorf("market maker number differs from the actual value; have %d, want %d", mmCount, depositCount)
+	}
+
+	reserveBalance := k.bankKeeper.GetAllBalances(ctx, types.DepositReserveAcc)
+	if !totalAmt.Empty() && reserveBalance.IsAllLT(totalAmt) {
+		return fmt.Errorf("DepositReserveAcc differs from the actual value; have %s, want %s", reserveBalance, totalAmt)
+	}
```

Comment:
I think we can avoid using `GetAllBalances` if `totalAmt` is empty.
```suggestion
	if !totalAmt.Empty() {
		reserveBalance := k.bankKeeper.GetAllBalances(ctx, types.DepositReserveAcc)
		if reserveBalance.IsAllLT(totalAmt) {
			return fmt.Errorf("DepositReserveAcc differs from the actual value; have %s, want %s", reserveBalance, totalAmt)
		}
	}
```